scripts/release.ts

```diff
@@ -0,0 +1,229 @@
+  await exec(`git reset HEAD~`);
+  await exec(`git checkout .`);
+  await exec(`git tag -d v${newVersion}`);
+}
+
+async function publish(tag: Tag): Promise<void> {
+  logger.log();
+  logger.log(`Publishing to ${tag}...`);
+  uiLine();
+  logger.log();
+
+  return new Promise((resolve, reject) => {
+    const childProcess = spawn('npm', ['publish', '--dry-run', '--tag', tag], {
+      stdio: 'inherit',
+      cwd: './dist',
+    });
+
+    childProcess.on('close', code => {
+      if (code !== EXIT_CODES.SUCCESS) {
+        reject();
+      } else {
+        resolve();
+      }
+    });
+  });
+}
+
+async function handler({
+  versionIncrement,
+  tag,
+}: ArgumentsCamelCase<ReleaseArguments>): Promise<void> {
+  setLogLevel(LOG_LEVEL.LOG);
+
+  const branch = await getGitBranch();
+
+  const isExperimental = tag === TAG.EXPERIMENTAL;
+
+  if (isExperimental && branch === MAIN_BRANCH) {
+    logger.error(
+      'Releases to experimental tag cannot be published from the main branch'
+    );
+    process.exit(EXIT_CODES.ERROR);
+  } else if (!isExperimental && branch !== MAIN_BRANCH) {
+    logger.error(
+      'Releases to latest and next tags can only be published from the main branch'
+    );
+    process.exit(EXIT_CODES.ERROR);
+  }
+
+  if (tag === TAG.LATEST && versionIncrement === INCREMENT.PRERELEASE) {
+    logger.error(
+      'Invalid release: cannot increment prerelease number on latest tag.'
+    );
+  }
+
+  const {
+    next: currentNextTag,
+    experimental: currentExperimentalTag,
+  } = await getDistTags();
+
+  if (!isExperimental && currentNextTag !== localVersion) {
+    logger.error(
+      `Local package.json version ${localVersion} is out of sync with published version ${currentNextTag}`
+    );
+    process.exit(EXIT_CODES.ERROR);
+  }
+
+  const currentVersion = isExperimental ? currentExperimentalTag : localVersion;
+  const prereleaseIdentifier = isExperimental ? 'experimental' : 'beta';
+  const incrementType =
+    tag === TAG.LATEST || versionIncrement === INCREMENT.PRERELEASE
+      ? versionIncrement
+      : (`pre${versionIncrement}` as const);
+
+  const newVersion = semver.inc(
+    currentVersion,
+    incrementType,
+    prereleaseIdentifier
+  );
+
+  if (!newVersion) {
+    logger.error('Error incrementing version.');
+    process.exit(EXIT_CODES.ERROR);
+  }
+
+  logger.log();
+  logger.log(`Current version: ${currentVersion}`);
+  logger.log(`New version to release: ${newVersion}`);
+
+  const shouldRelease = await confirmPrompt(
+    `Release version ${newVersion} on tag ${tag}?`
+  );
+
+  if (!shouldRelease) {
+    process.exit(EXIT_CODES.SUCCESS);
+  }
+
+  logger.log();
+  logger.log(`Updating version to ${newVersion}...`);
+  await exec(`yarn version --new-version ${newVersion}`);
+  logger.success('Version updated successfully');
+
+  logger.log();
+  logger.log('Creating a new build...');
+  await build();
+  logger.success('Build successful');
+
+  try {
+    await publish(tag);
```

Comment:
I wonder if we should try to enforce the QA --> PROD release flow here. We could check to see if the user has released these changes to the next branch, and encourage them to do it if they haven't.

For example, if the user attempts to increment from version 1.0.0 to version 1.0.1, we could first check to see if version 1.0.1-beta.<n> exists (where n doesn't matter). If it exists, then we can assume that the current local changes have already been released to the QA (next) tag. In that case, We can safely publish version 1.0.1, and we can still publish version 1.0.1-beta.<n+1>

Question: What should the version in the CLI's package.json be after this? 1.0.1 or 1.0.1-beta<n+1>?

If it doesn't exist (i.e. maybe only 1.0.0-beta.1 exists), we could prompt something like: `It looks like these changes have not been released to the next tag yet, we encourage you to release changes to the next tag before releasing them to the latest tag. Would you like to continue anyways?`. If they choose yes (Which IMO "no" should be the default option), we can continue to "dangerously" publish 1.0.1, and then we can also publish 1.0.1-beta.0 just to catch the next version up.

Here's an example of what the flow would look like:
1. [just to establish initial state] `yarn publish -v=major -tag=latest` --> publish version 1.0.0
2. `yarn publish -v=patch -tag=latest` --> We prompt to release these changes to next tag first, dev agrees
3. `yarn publish -v=patch -tag=next` --> publish version 1.0.1-beta.0
4. `yarn publish -v=patch -tag=latest` --> publish versions 1.0.1 and 1.0.1-beta.1

Alternatively...
1. [just to establish initial state] `yarn publish -v=major -tag=latest` --> publish version 1.0.0
2. `yarn publish -v=patch -tag=latest` --> We prompt to release these changes to next tag first, dev does not agree and continues to bypass QA (next) --> publish versions 1.0.1 and 1.0.1-beta.0